baybe/surrogates/utils.py

```diff
@@ -57,156 +55,101 @@ def _prepare_targets(y: Tensor) -> Tensor:
     return y.to(DTypeFloatTorch)
 
 
-def catch_constant_targets(model_cls: type[Surrogate]):
-    """Wrap a ``Surrogate`` class that cannot handle constant training target values.
+def catch_constant_targets(cls: type[Surrogate], std_threshold: float = 1e-6):
+    """Make a ``Surrogate`` class robustly handle constant training targets.
+
+    More specifically, "constant training targets" can mean either of:
+        * The standard deviation of the training targets is below the given threshold.
+        * There is only one target and the standard deviation cannot even be computed.
 
-    In the wrapped class, these cases are handled by a separate model type.
+    The modified class handles the above cases separately from "regular operation"
+    by resorting to a :class:`baybe.surrogates.naive.MeanPredictionSurrogate`.
 
     Args:
-        model_cls: A ``Surrogate`` class that should be wrapped.
+        cls: The :class:`baybe.surrogates.base.Surrogate` to be augmented.
+        std_threshold: The standard deviation threshold below which operation is
+            switched to the alternative model.
 
     Returns:
-        A wrapped version of the class.
+        The modified class.
     """
+    from baybe.surrogates.naive import MeanPredictionSurrogate
 
-    class SplitModel(*model_cls.__bases__):
-        """The class that is used for wrapping.
-
-        It applies a separate recommender for cases where the training
-        targets are all constant and no variance can be estimated.
-
-        It stores an instance of the underlying model class.
-        """
-
-        # The posterior mode is chosen to match that of the wrapped model class
-        joint_posterior: ClassVar[bool] = model_cls.joint_posterior
-        # See base class.
-
-        def __init__(self, *args, **kwargs):
-            super().__init__()
-            self.model = model_cls(*args, **kwargs)
-            self.__class__.__name__ = self.model.__class__.__name__
-
-        def _posterior(self, candidates: Tensor) -> tuple[Tensor, Tensor]:
-            """Call the posterior function of the internal model instance."""
-            import torch
+    # References to original methods
+    _fit_original = cls._fit
+    _posterior_original = cls._posterior
 
-            mean, var = self.model._posterior(candidates)
+    def _posterior_new(self, candidates: Tensor) -> tuple[Tensor, Tensor]:
+        # Alternative model fallback
+        if isinstance(self._model, MeanPredictionSurrogate):
```

Comment:
This was a really hard question and I needed to double-check with the debugger to be able to answer, in fact 🍔 (random emoji to express my feelings)

I think I understand your gutfeeling here and was wondering the same once you asked, but then I realized that what you probably meant is: why not call `self._posterior` in both cases? Because `self._model` actually refers to the "external" model (such as `ARDRegression`), which doesn't offer the `posterior` interface.

Anyway, I **guess** both calls could actually replace with a single `self._posterior` call, which in fact would even have the effect that the entire method replacement could vanish. However, I'm not 100% if this is safe to do, because it always comes with the implicit assumption that the original `_posterior` call **only** queries the internal model and does nothing else (because this is what NEEDS to be happen for the `MeanPredictionSurrogate`). Currently, this seems to be the case, but we actually don't impose this anywhere! The base class only declares the abstract `_posterior` method and it's completely up to the subclasses how the populate it. Therefore, I feel that we can silently run into errors in the future.

While the current implementation **might** invoke one additional unnecessary call, it's at least bullet-proof and unmistakably clear in what it does: if now wrapper is active, call the original method, otherwise precisely do the necessary logic defined inside the wrapper body.

Any thoughts? Also from @Scienfitz?

Reply:
Sorry, but I do not understand your explanation. For me, it seems like there are two cases.
1. `self._model` is a `MeanPredictionSurrogate`: In this case, I return `self._model._posterior(candidates)`.
2. This is not the case: Then, I return `_posterior_original(self, candidates)` which is just the return value of `_posterior(candidates)` that the wrapped object belongs to - which is `self._model._posterior`. This is at least how I understand the code and the logic here.  
And in the first case, `self._model` does also not refer to the "external" model since it then refers to `MeanPredictionSurrogate` which is our own model. So I unfortunately do not understand the point here ❓

Reply:
1. correct
2. I think you are slightly off here: it doesn't point to `self._model._posterior`, but `self._posterior`, which is potentially where the confusion lies. But let me adjust a few things in the decorator and along the way rename the field from `_model` to `_constant_target_model` (since we'll probably remove the `_model` attribute from the baseclass anyway). Once you see the changed name, you'll no longer see the (wrong) association between the two cases, I think. Or perhaps I'm going to be lost as well as soon as I start working on it again :D

Reply:
Hi @AVHopp, please have another look now. The new version makes it explicitly clear where stuff belongs to and I hope it should also answer your question? Let me know if not ...